File: day_18/p1.py
from typing import Dict, Generator, List, Tuple
from collections import deque

NROWS = 71
NCOLS = 71


def parse(fname: str) -> List[Tuple[int, ...]]:
    lines = []

    with open(fname, 'r') as f:
        lines = [tuple(map(int, line.split(',')[::-1]))
                for line in f.read().splitlines()]
    
    '''
    Each byte position is given as an X,Y coordinate, where
    - X is the distance from the left edge of your memory space
    - Y is the distance from the top edge of your memory space.

    Feels more natural the other way around, hence I flipped Xs and Ys
    '''
    
    return lines

def move_gen(cx: int, cy: int) -> Generator[Tuple[int, int], None, None]:
    yield -1,  0
    yield  1,  0
    yield  0,  1
    yield  0, -1

def dijkstra(grid: List[List[str]],
             start: Tuple[int, int],
             end: Tuple[int, int],
) -> int:
    x, y = start
    visited: Dict[Tuple[int, int], int] = {}
    steps = []
    step = 0
    
    q = deque([( (start), step) ])
    while q:
        (cx, cy), step = q.popleft()

        if (cx, cy) == end:
            print("here")
            steps.append(step)
            continue

        state = (cx, cy)
        if state in visited:
            continue
        visited[state] = step
        
        for dx, dy in move_gen(cx, cy):
            nx = cx + dx
            ny = cy + dy
            if nx in range(NROWS) and ny in range(NCOLS) and grid[nx][ny] != '#':
                q.append(( (nx, ny), step + 1 ))

    return min(steps)

def compute(coords: List[Tuple[int]], nbytes: int) -> int:
    grid: List[List[str]] = []
    for i in range(NROWS):
        grid.append([])

        for j in range(NCOLS):
            if (i, j) not in set(coords[:nbytes]):
                grid[i].append('.')
            else:
                grid[i].append('#')

    return dijkstra(grid, (0, 0), (NROWS-1, NCOLS-1))

def main():
    coordinates = parse("input.txt")
    print(compute(coordinates, 1024))


if __name__ == "__main__":
    main()

File: day_18/p1.cpp
#include "pypp.hpp" // https://github.com/A-Bergonzoli/pypp/tree/main

#include <iostream>

constexpr auto NROWS { 71 };
constexpr auto NCOLS { 71 };

struct Move {
    int dx;
    int dy;
    grid::Direction hdg;
};

std::vector<std::pair<int, int>> Parse(const std::string& fname)
{
    std::vector<std::pair<int, int>> coordinates {};
    auto lines = pypp::splitFileLines(fname);

    for (auto line : lines) {
        const auto y = pypp::splitThenGetAt(line, ',', 0);
        const auto x = pypp::splitThenGetAt(line, ',', 1);
        coordinates.push_back({ std::stoi(x), std::stoi(y) });
    }

    return coordinates;
}

std::array<std::pair<Move, int>, 4U> Generator(grid::Direction hdg)
{
    return {
        std::make_pair(Move { -1, 0, grid::Direction::NONE }, 1),
        std::make_pair(Move { +1, 0, grid::Direction::NONE }, 1),
        std::make_pair(Move { 0, +1, grid::Direction::NONE }, 1),
        std::make_pair(Move { 0, -1, grid::Direction::NONE }, 1)
    };
}

int Compute(const std::vector<std::pair<int, int>>& coords, int num_bytes)
{
    strings grid(NROWS, std::string(NCOLS, '.'));

    std::unordered_set<std::pair<int, int>, collections::TupleHash> coords_to_consider {};
    for (int i { 0 }; i < num_bytes; ++i)
        coords_to_consider.insert({ coords[i].first, coords[i].second });

    for (std::size_t i { 0U }; i < NROWS; ++i)
        for (std::size_t j { 0U }; j < NCOLS; ++j)
            if (coords_to_consider.contains({ i, j }))
                grid[i][j] = '#';

    const auto start = std::make_tuple(0, 0, grid::Direction::RIGHT);
    const auto end = std::make_pair(NROWS - 1, NCOLS - 1);

    auto [min_score, _] = algorithms::dijkstra(grid, start, end, Generator);
    return min_score;
}

int main()
{
    const auto coords = Parse("input.txt");
    std::cout << Compute(coords, 1024) << std::endl;

    return 0;
}
